src/test/mds/TestQuiesceDb.cc

```diff
@@ -0,0 +1,824 @@
+    r.set_id = "set2";
+    r.timeout = sec(60);
+    r.expiration = sec(60);
+    r.include_roots({ "root2", "root3"});
+  }));
+
+  std::queue<TestRequestContext*> outstanding_awaits;
+  std::queue<TestRequestContext*> pending_requests;
+
+  // let's have several awaits pending
+  for(auto&& set_id: {"set1", "set2"}) {
+    for (int i=0; i<2; i++) {
+      outstanding_awaits.emplace(&start_request([set_id](auto&r) {
+        r.set_id = set_id;
+        r.await = sec(100);
+      }));
+      EXPECT_EQ(NA(), outstanding_awaits.front()->check_result());
+    }
+  }
+
+  // flush the pending requests by running a simple query
+  EXPECT_EQ(OK(), run_request([](auto& r) { r.query("set1"); }));
+
+  ASSERT_EQ(outstanding_awaits.size(), managers.at(0)->internal_awaits().size());
+
+  std::mutex agent_mutex;
+  std::condition_variable agent_cond;
+  bool callback_reached = false;
+
+  // block the db thread with a malicious agent callback
+  managers.at(0)->reset_agent_callback([&](auto& map) {
+    std::unique_lock l(agent_mutex);
+    callback_reached = true;
+    agent_cond.notify_all();
+    l.unlock();
+    std::this_thread::sleep_for(sec(0.1));
+    return false;
+  });
+
+  {
+    std::unique_lock l(agent_mutex);
+    agent_cond.wait(l, [&]{return callback_reached;});
+  }
+
+  // now that the db thread is sleeping we can pile up some pending requests
+  pending_requests.emplace(&start_request([](auto& r) {
+    r.set_id = "set3";
+    r.include_roots({"root4"});
+  }));
+  EXPECT_EQ(NA(), pending_requests.front()->check_result());
+
+  pending_requests.emplace(&start_request([](auto& r) {
+    r.set_id = "set4";
+    r.include_roots({"root5"});
+  }));
+  EXPECT_EQ(NA(), pending_requests.front()->check_result());
+
+  pending_requests.emplace(&start_request([](auto& r) {
+    r.set_id = "set1";
+    r.await = sec(100);
+  }));
+  EXPECT_EQ(NA(), pending_requests.front()->check_result());
+
+  ASSERT_EQ(managers.at(0)->internal_pending_requests().size(), pending_requests.size());
+
+  // reset the membership of the manager
+  // this will block until the db thread exits
+  managers.at(0)->update_membership({});
+
+  // as of now all requests must have finished
+  while(!outstanding_awaits.empty()) {
+    auto& r = *outstanding_awaits.front();
+    EXPECT_EQ(ERR(EINPROGRESS), r.check_result());
+    outstanding_awaits.pop();
+  }
+
+  while (!pending_requests.empty()) {
+    auto& r = *pending_requests.front();
+    EXPECT_EQ(ERR(EPERM), r.check_result());
+    pending_requests.pop();
+  }
+}
+
+TEST_F(QuiesceDbTest, MultiRankQuiesce)
+{
+  ASSERT_NO_FATAL_FAILURE(configure_cluster({ 0, 1, 2 }));
+
+  std::vector<TestRequestContext*> awaits;
+
+  // create and await for several sets
+  for (auto&& set_id: {"set1", "set2", "set3", "set4"}) {
+    awaits.emplace_back(&start_request([set_id](auto& r) {
+      r.set_id = set_id;
+      r.timeout = sec(100);
+      if (strstr(set_id, "set4")) {
+        r.expiration = sec(100);
```

Comment:
So here you expect set[123] to expire immediately. That could be a separate test.

Reply:
correct. I use multiple sets to validate different flows concurrently. I agree that separate tests would look cleaner, but at this point I tried to achieve max coverage with minimal time/code